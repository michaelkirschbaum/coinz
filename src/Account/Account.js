import React from 'react'

const Account = () => (
  <div>
    Balance
  </div>
)

export default Account