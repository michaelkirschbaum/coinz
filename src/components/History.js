/** @jsx jsx */
import React, { useRef, useEffect, useContext } from 'react'
import * as d3 from 'd3'
import { UserContext } from '../App'
import PropTypes from 'prop-types'
import { css, jsx } from '@emotion/core'

const History = ({ transactions }) => {
  const ref = useRef(null)
  useEffect(() => {
    if (transactions) {
      // clear previous
      d3.selectAll("svg > *").remove();

      drawChart()
    }
  })
  const user = useContext(UserContext)

  const drawChart = () => {
    const xScale = d3.scaleTime()
      .domain(d3.extent(transactions, (d) => new Date(d.timestamp)))
      .range([0, 835.422])

    const yScale = d3.scaleLinear()
      .domain([0, 100])
      .range([441, 0])

    // create graph container
    const svg = d3.select(ref.current)
      .append('g')
      .attr("transform", "translate(" + 20 + "," + 20 + ")");

    // x axis
    svg.append('g')
      .attr('transform', `translate(0, 441)`)
      .call(d3.axisBottom(xScale))

    // y axis
    svg.append('g')
      .call(d3.axisLeft(yScale))

    // add balance to data
    let balance = 0
    transactions.map((entry) => {
      if (entry.toAddress === user)
        balance += parseFloat(entry.amount)
      else
        balance -= parseFloat(entry.amount)

      entry.balance = balance
    })

    // generate line
    const line = d3.line()
      .x(d => xScale(new Date(d.timestamp)))
      .y(d => yScale(d.balance))

    // bind data
    svg.append('path')
      .datum(transactions)
      .style('stroke','black')
      .style('stroke-width', 1)
      .style('fill', 'none')
      .attr('d', line)

    // add data points
    svg.selectAll('circle')
      .data(transactions)
      .enter()
      .append('circle')
      .attr('cx', d => xScale(new Date(d.timestamp)))
      .attr('cy', d => yScale(d.balance))
      .attr('r', 4)
      .style('fill', '#EF7293')
      .style('stroke', 'black')
      .style('stroke-width', 1)
  }

  return (
    <svg
      css={css`
        height: 100vh;
        width: 100%;
        border: 1px solid lightgrey;
      `}
      ref={ref}
    />
  )
}

History.propTypes = {
  transactions: PropTypes.arrayOf(
    PropTypes.shape({
      timestamp: PropTypes.string,
      toAddress: PropTypes.string,
      fromAddress: PropTypes.string,
      amount: PropTypes.string,
    })
  )
}

export default History