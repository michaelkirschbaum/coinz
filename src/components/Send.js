/** @jsx jsx */
import React from 'react'
import { css, jsx } from '@emotion/core'

class Send = () => (
  <div
    css={css`
      display: grid;
    `}
  >
    <div>Send Jobcoin</div>
    <div>
      Destination Address
      <input />
    </div>
    <div>
      Amount
      <input />
    </div>
    <button>Send</button>
  </div>
)

export default Send