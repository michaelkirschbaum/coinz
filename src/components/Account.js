/** @jsx jsx */
import React, { useState, useEffect, useContext } from 'react'
import { UserContext } from '../App'
import Send from 'components/Send'
import History from 'components/History'
import { css, jsx } from '@emotion/core'

const Account = () => {
  const [data, setData] = useState({})
  const [error, setError] = useState(false)
  const user = useContext(UserContext)
  useEffect(() => {
    setError(false)

    {/* fetch jobcoin balance */}
    const fetchData = async () => {
      try {
        const result = await fetch('/api/addresses/' + user)
        const data = await result.json()
        setData(data)
      } catch (error) {
        setError(true)
      }
    }
    fetchData()
  }, [])

  return (
    <div css={css`
      display: flex;
      flex-direction: row;
      margin: 25px 25px 25px 25px;
    `}>
      <div css={css`
        width: 300px;
        margin-right: 25px;
      `}>
        <div css={css`
          border: 1px solid lightgrey;
          margin-bottom: 10px;
          padding: 0 10px 0 10px;
        `}>
          <h4>Jobcoin Balance</h4>
          {error
            ? <p>unable to get balance...</p>
            : <p>{data.balance}</p>}
        </div>
        <Send />
      </div>
      <History transactions={data.transactions} />
    </div>
  )
}

export default Account