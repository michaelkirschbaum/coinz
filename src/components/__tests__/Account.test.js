import React from 'react'
import { shallow } from 'enzyme'
import Account from 'components/Account'
import Send from 'components/Send'
import History from 'components/History'

describe('Account', () => {
  let account
  beforeEach(() => {
    account = shallow(
      <Account accountID={'Alice'} />,
    )
  })

  it('should match the snapshot', () => {
    expect(account).toMatchSnapshot()
  })

  it('should render balance', () => {
    global.fetch = jest.fn().mockImplementationOnce(() => Promise.resolve({
      json: () => Promise.resolve({ data: { balance: "10" } })
    }))

    account = shallow(
      <Account accountID = {'Alice'}/>,
    )
    expect(account.text()).toContain('10')

    global.fetch.mockClear()
  })

  it('should update balance', () => {

  })

  it('should show error when balance request fails', () => {
    global.fetch = jest.fn().mockImplementationOnce(() => Promise.reject(
      new Error('Fetch failed')
    ))

    account = shallow(
      <Account accountID = {'Alice'}/>,
    )
    expect(account.text()).toContain('unable to get balance')

    global.fetch.mockClear()
  })

  it('should render Send widget', () => {
    expect(account.find(Send)).toHaveLength(1)
  })

  it('should render history graph', () => {
    expect(account.find(History)).toHaveLength(1)
  })
})