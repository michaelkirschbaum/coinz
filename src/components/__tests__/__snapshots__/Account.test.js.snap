// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account should match the snapshot 1`] = `
<Fragment>
  <EmotionCssPropInternal
    __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Account"
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "y8ucs5",
        "next": undefined,
        "styles": "

        ",
      }
    }
  >
    <div>
      Balance
    </div>
    <div />
  </EmotionCssPropInternal>
  <Send />
  <History />
</Fragment>
`;
